### PR TITLE
[docs] Fix DataGrid[Pro/Premium] reference links

### PR DESCRIPTION
**Pull request by michelengelen**
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

--- 

This adds Demo and API links to the export for `DataGrid`, `DataGridPro` and `DataGridPremium`

It's part of a series to solve #9226

### CODE REVIEW

File: packages/grid/x-data-grid-premium/src/DataGridPremium/DataGridPremium.tsx [hunk 1]
```diff
@@ -60,6 +60,16 @@ interface DataGridPremiumComponent {
   propTypes?: any;
 }
 
+/**
+ *
+ * Demos:
+ *
+ * - [DataGridPremium](https://mui.com/x/react-data-grid/demo/)
```
Comment (michelengelen) [inline packages/grid/x-data-grid-premium/src/DataGridPremium/DataGridPremium.tsx:67]: little question here: the API pages link to the overview page and not the demo page directly. Should we follow this approach with these links as well, or would it be better to directly link to the Demo page?

File: packages/grid/x-data-grid-premium/src/DataGridPremium/DataGridPremium.tsx [hunk 2]
```diff
@@ -60,6 +60,16 @@ interface DataGridPremiumComponent {
   propTypes?: any;
 }
 
+/**
+ *
```
Comment (romgrk) [inline packages/grid/x-data-grid-premium/src/DataGridPremium/DataGridPremium.tsx:64]: First newline in these comments is superfluous, we can keep this more compact.

Comment (romgrk) [reply packages/grid/x-data-grid-premium/src/DataGridPremium/DataGridPremium.tsx:64]: And I think maybe there is a JSDoc annotation for these? Not sure, to be verified.

Comment (michelengelen) [reply packages/grid/x-data-grid-premium/src/DataGridPremium/DataGridPremium.tsx:64]: There is one, but generally both work just fine in this case. Is it worth the effort of changing those? I cannot find any benefit from JSDoc annotation over MD style.

File: packages/grid/x-data-grid-premium/src/DataGridPremium/DataGridPremium.tsx (hunk 1 above)
Comment (alexfauquette) [reply packages/grid/x-data-grid-premium/src/DataGridPremium/DataGridPremium.tsx:67]: Yes, the hash of the link in API page is broken, it should redirect to https://mui.com/x/react-data-grid/#commercial-versions I assume the choice of the section "Commercial versions" is to ensure people discovering it knows about the different plans.

For me linking to demo looks good :+1:

File: packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterForm.tsx [hunk 3]
```diff
@@ -651,4 +651,16 @@ GridFilterForm.propTypes = {
   valueInputProps: PropTypes.any,
 } as any;
 
+/**
+ *
+ * Demos:
+ *
+ * - [DataGrid](https://mui.com/x/react-data-grid/)
+ * - [DataGridPro](https://mui.com/x/react-data-grid/#pro-plan)
+ * - [DataGridPremium](https://mui.com/x/react-data-grid/#premium-plan)
```
Comment (alexfauquette) [inline packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterForm.tsx:660]: What about something more filtering focus?

```suggestion
 * - [Filtering - overview](https://mui.com/x/react-data-grid/filtering/)
```

**Review by alexfauquette — APPROVED** [review]

File: packages/grid/x-data-grid/src/components/toolbar/GridToolbarQuickFilter.tsx [hunk 4]
```diff
@@ -182,4 +182,11 @@ GridToolbarQuickFilter.propTypes = {
   quickFilterParser: PropTypes.func,
 } as any;
 
+/**
+ * Demos:
+ * - [Filtering - overview](https://mui.com/x/react-data-grid/filtering/)
```
Comment (MBilalShafi) [inline packages/grid/x-data-grid/src/components/toolbar/GridToolbarQuickFilter.tsx:187]: Will direct access to the quick filtering page be more meaningful?
```suggestion
 * - [Filtering - quick filter](https://mui.com/x/react-data-grid/filtering/quick-filter/)
```

**Review by MBilalShafi — APPROVED** [review]
Nice improvement 🎉

File: packages/grid/x-data-grid/src/components/toolbar/GridToolbarQuickFilter.tsx (hunk 4 above)
Comment (michelengelen) [reply packages/grid/x-data-grid/src/components/toolbar/GridToolbarQuickFilter.tsx:187]: Nice catch ... IMO there is no harm in adding this as an additional ressource, so I just did an additional line linking that detailed demo page 💯

Comment (MBilalShafi) [reply packages/grid/x-data-grid/src/components/toolbar/GridToolbarQuickFilter.tsx:187]: That's even better 👍

**Pull request merged**